include/crab/analysis/inter/top_down_inter_analyzer.hpp

```diff
@@ -1032,6 +1032,104 @@ class top_down_inter_transformer final
     // crab::CrabStats::count("Interprocedural.num_calling_contexts");
   }
 
+  /// @brief group regions variables in parameter lists based on target's dsa
+  /// intrinsics
+  /// @param tgt_cfg the cfg for target function
+  /// @param src_params the paramters appeared in the source function
+  /// @param tgt_params the paramters appeared in the target function
+  /// @param src_cls the equivalence classes for grouping source regions
+  /// @param tgt_cls the equivalence classes for grouping target regions
+  static void group_regions_by_dsa_intrinsics(
+      const cfg_t &tgt_cfg, const std::vector<variable_t> &src_params,
+      const std::vector<variable_t> &tgt_params,
+      std::vector<std::vector<variable_t>> &src_cls,
+      std::vector<std::vector<variable_t>> &tgt_cls) {
+
+    using variable_or_constant_t =
+        typename intrinsic_statement_t::variable_or_constant_t;
+    // Helper lambda functions for logging
+    auto print_var_vector = [](crab::crab_os &o,
+                               const std::vector<variable_t> &vec) {
+      typename std::vector<variable_t>::const_iterator it;
+      o << "[";
+      for (it = vec.begin(); it != vec.end(); it++) {
+        if (it != vec.begin())
+          o << ",";
+        o << (*it);
+      }
+      o << "]";
+    };
+    auto print_var_or_const_vector =
+        [](crab::crab_os &o, const std::vector<variable_or_constant_t> &vec) {
+          typename std::vector<variable_or_constant_t>::const_iterator it;
+          o << "[";
+          for (it = vec.begin(); it != vec.end(); it++) {
+            if (it != vec.begin())
+              o << ",";
+            o << (*it);
+          }
+          o << "]";
+        };
+
+    CRAB_LOG("inter-group", errs() << "Ins: ";
+             print_var_vector(errs(), src_params); errs() << "\n";);
+    CRAB_LOG("inter-group", errs() << "Outs: ";
+             print_var_vector(errs(), tgt_params); errs() << "\n";);
+
+    // Step 1. get all dsa intrinsic statements
+    const basic_block_t &entry_bb = tgt_cfg.get_node(tgt_cfg.entry());
+    std::vector<intrinsic_statement_t> intrinsic_stmts;
+    intrinsic_stmts.reserve(entry_bb.size());
+    for (auto const &stmt : entry_bb) {
+      if (stmt.is_intrinsic()) {
+        auto intrinsic_stmt = dynamic_cast<const intrinsic_statement_t &>(stmt);
+        if (intrinsic_stmt.get_intrinsic_name() ==
+            "regions_from_memory_object") {
+          intrinsic_stmts.push_back(intrinsic_stmt);
+        }
+      }
+    }
+
+    // Step 2. construct the equivalence classes for grouping input/output
+    // regions Note that, the order of the output regions in the parameter list
+    // is the same as the output list in the dsa intrinsic
+    // The worst case of the following loop is O(m * n) where m is the number of
+    // parameters and n is the number of dsa intrinsic statements
+    unsigned int index = 0, param_lst_sz = src_params.size();
```

Comment:
if this is just executed once (before the analysis starts) performance might not be an issue but the current algorithm is far from ideal. As you mention, for each argument we go over all intrinsics. This is completely unnecessary.  I would process the intrinsics only once and put them in some sort of union-find. Then, for each argument in the caller you get the argument in the callee and then you can  query the union-find to see all the elements in the equivalence class. Note that you don't a super fancy union-find implementation. You don't even need to write a separate union-find class. Just have an internal map from a variable to its parent and do standard path-compression when you do the queries.